CARDS-2288: Selectable Areas Question

acrowthe opened the pull request:
- Add new selectable areas question type
- New question type is extendable with different images and area maps
  - Add selectable body part extension for SPARC questionnaires

A new questionnaire has been added for the test runmode. It can be tested using `./start_cards.sh --test --dev` and creating a new `SelectableAreasTest` form

![image](https://github.com/data-team-uhn/cards/assets/20056751/1f2fe9bb-edc6-413e-b481-4f1939609fac)

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Add new selectable areas question type
- New question type is extendable with different images and area maps
  - Add selectable body part extension for SPARC questionnaires

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Add configuration for stroke color and highlight color
- Whitespace and unused import improvement

acrowthe commented:
Conditional sections referencing body questions do not update their visibility unless the form is reloaded

acrowthe commented:
Added options for stroke color and different highlight vs. hover colors:

https://github.com/data-team-uhn/cards/assets/20056751/48370d31-b892-4253-b838-566594ba873d

Created temporary branch with a more complicated test/demo form, also accessible via the test run mode:
https://github.com/data-team-uhn/cards/tree/CARDS-2288-demo

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Fix conditional sections referencing selectable area questions

marta- commented on modules/data-entry/src/main/frontend/package.json:
```diff
@@ -56,6 +56,7 @@
     "uuid": "9.0.0",
-    "@uiw/react-md-editor": "3.20.2"
+    "@uiw/react-md-editor": "3.20.2",
+    "react-img-mapper": "1.5.1"
```
Let's move this up with the other `react-` deps

marta- commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBodyMap.json:
```diff
@@ -0,0 +1,537 @@
+    "title": "Face",
+    "value": "face",
+    "shape": "poly",
+    "name": "1",
```
why is `name` a number?

marta- commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBodyMap.json:
```diff
@@ -0,0 +1,537 @@
+    ]
+  },
+  {
+    "id": "",
```
why are ids empty starting here?

marta- commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBodyMap.json:
```diff
@@ -0,0 +1,537 @@
+    "title": "Right Chest/Breast",
+    "value": "rightchest",
+    "shape": "rect",
+    "name": "",
```
why are names empty starting here?

marta- commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBodyMap.json:
```diff
@@ -0,0 +1,537 @@
+    "value": "face",
+    "shape": "poly",
+    "name": "1",
+    "fillColor": "#b5b5b5b5",
```
I like the fact that we can specify the fill color, but I would like it more if there was a default that this field could optionally override.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,252 @@
+  let onAreaClicked = (area, index) => {
+    let clickedEntry = [area.title, area.value];
+
+    if (maxAnswers == 1) {
+      // Single select: Just set selection to the most recent value
+      selections.pop();
+      selections.push(clickedEntry);
+    } else {
+      // Multi select: toggle the values' selection state
+      let entryIndex = -1;
+      selections.forEach((selectionValue, selectionIndex) => {
+        if (selectionValue[VALUE_POS] === clickedEntry[VALUE_POS]) {
+          entryIndex = selectionIndex;
+        }
+      })
+      if(entryIndex == -1) {
+        selections.push(clickedEntry);
+      } else {
+        selections.splice(entryIndex, 1);
+      }
+    }
+
+    let newSelections = selections.concat()
+    setSelections(newSelections)
```
This might end up buggy and unpredictable.

It's recommended to do all the processing within the state setter, something like this (not tested):

```suggestion
    setSelections(oldSel => {
      if (maxAnswers == 1) {
        // Single select: Just set selection to the most recent value
        return [clickedEntry];
      }
      // Multi select: toggle the values' selection state
      let newSel = oldSel.slice();
      let entryIndex = -1;
      newSel.forEach((selectionValue, selectionIndex) => {
        if (selectionValue[VALUE_POS] === clickedEntry[VALUE_POS]) {
          entryIndex = selectionIndex;
        }
      })
      if(entryIndex == -1) {
        newSel.push(clickedEntry);
      } else {
        newSel.splice(entryIndex, 1);
      }
      return newSel;
    });
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SelectableAreaQuestion.jsx:
```diff
@@ -0,0 +1,252 @@
+    .map((item, index) => [Array.of(existingAnswer[1].displayedValue).flat()[index], item]);
+
+
+  const [ selections, setSelections ] = useState(initialSelection);
```
`selections` (plural) sounds awkward to me, could we name it `selection` (singular) instead? "Selection" already means **several** chosen things.

marta- commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBodyMap.json:
```diff
@@ -0,0 +1,537 @@
+    ]
+  },
+  {
+    "id": "",
+    "title": "Abdomen",
+    "value": "abdomen",
+    "shape": "poly",
+    "name": "",
+    "fillColor": "#b5b5b5b5",
+    "coords": [
+      260, 371,
+      445, 371,
+      445, 530,
+      380, 590,
+      325, 590,
+      260, 530
+    ]
+  },
+  {
+    "id": "",
+    "title": "Pelvis",
+    "value": "pelvis",
+    "shape": "poly",
+    "name": "",
+    "fillColor": "#b5b5b5b5",
+    "coords": [
```
I find it strange that Abdomen and Pelvis are listed without sides.

acrowthe replied:
I did not add Abdomen or Pelvis sides but with the new AnswerOptions implementation it will be easy to add in the future if needed without impacting existing forms

marta- commented on modules/data-entry/src/main/media/SLING-INF/content/libs/cards/resources/media/dataEntry/fullBodyMap.json:
```diff
@@ -0,0 +1,537 @@
+  },
+  {
+    "id": "",
+    "title": "Left Upper Arm",
```
I find it a bit strange that I can't select the arms and legs on the back image. I guess it may be ok for SPARC as it complies to their specs, but in general I think people should be able to say the back of their knee or the sole of their foot hurts. Thoughts?

acrowthe replied:
Added areas to select arms and legs from the back view now: Selecting either front or back will highlight both and only be saved as a single value so UX is improved but outputs in the current use case are the same

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Change mouse cursor to pointer when hovering over an area

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Add tooltip when areas hovered

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Improve color configuration and add default colors
- Adjust text selection display tpo float beside map when space allows
- Improve code readability and prepare to make better use of useEffect

acrowthe committed:
CARDS-2288: Selectable Areas Question
Address review comments
- Clean up unneeded map fields
- Fix a couple typos in map fields
- Refactor selection name and assignment

acrowthe committed:
CARDS-2288: Selectable Areas Question
Update Selectable Areas Question to support AnswerOptions
- Add additional areas to FullBody map to provide additional options for left/right/combined
- Add additional areas to FullBody map to make arms and legs selectable on the back image
- Only display selectable areas that are listed as an answer option
- Use AnswerOptions for displayed values improving handling in view mode

sashaandjic reviewed (approved):
tested - looks and behaves good

Question: I wonder if it would look good to we leave the body displayed in the view mode as well.

sashaandjic commented:
`Buttock` or 'Buttocks` here? There is no option for left and right.
![image](https://github.com/data-team-uhn/cards/assets/23243244/8668ce05-e4dd-4b38-b1c6-bf9899aef35e)

marta- commented:
> `Buttock` or 'Buttocks` here? There is no option for left and right. 

Agreed it should be plural if it's both at the same time.

marta- commented:
**UX issue**:  when maxAnswers is 1 (single select), clicking on a different area doesn't remove the shading from the previously selected area until the cursor leaves the newly selected one.

marta- commented:
Also, the picture with the selected areas should be visible in view mode.

marta- commented:
**Bug**: WSORD when trying to update the SelectableAreasTest questionnaire in the questionnaire designer.

sashaandjic commented:
> **UX issue**: when maxAnswers is 1 (single select), clicking on a different area doesn't immediately remove the shading from the previously selected area; only when the cursor leaves the newly selected one is the "selected" shading reassigned.

Also, when clicking on a body part, the `light gray` should become `dark blue`, and 'dark blue` should become `gray`. For example you can click on gray many times without knowing if it is selected or not. Same for selected body part 'dark blue`, yo can select/deselect without the body part changing the color.

sashaandjic commented:
Screenshot for WSORD when going in the design mode of the test form that @marta-  reported:
![image](https://github.com/data-team-uhn/cards/assets/23243244/6d9039c8-685d-434b-9c75-9b5478a39bf6)

acrowthe committed:
CARDS-2288: Selectable Areas Question
- Add notApplicable support via checkbox above question if required
- Adjust areas to match required areas
- Add additional test cases for new areas and options

acrowthe commented:
Replaced by new branch and pull request here: https://github.com/data-team-uhn/cards/pull/1691
